mixins/localize-mixin.md

```diff
@@ -0,0 +1,294 @@
+
+The user's language, timezone and any D2L locale overrides are automatically fetched by `LocalizeMixin` from the `<html>` element's `lang`, `data-timezone` and `data-intl-overrides` attributes respectively.
+
+## Usage
+
+To use `LocalizeMixin`, apply the mixin then implement the following functions (examples in "Language Resources"):
+* `getLanguage(langs)`: returns the page's language based upon `langs`, an array of possible languages determined by `LocalizeMixin` (e.g., `['ar-dz', 'ar', 'en']`)
+* `async getLangResources(lang)`: retrieves the available language resources for `lang`
+
+Consume your web component in a page which has the lang attribute set on the <html> element:
+```html
+<html lang="ar">
+	<head>
+		<script type="module">
+			import '../my-element.js';
+		</script>
+	</head>
+	<body>
+		<my-elem></my-elem>
+	</body>
+</html>
+```
+
+If the language of the page changes (via an update to the `lang` attribute on `<html>`), the mixin will automatically detect the change and re-render the web component. It will fire the event `d2l-localize-behavior-language-changed` when this occurs.
+
+## Language Resources
+
+* Always provide entries for base languages (e.g. "en", "fr", "pt") so that if the user is using a regional language (e.g. "en-gb", "fr-ca", "pt-br") which is missing, it can fall back to the base language.
+* If there's no entry for a particular language, and no base language, the value of `data-lang-default` on the `<html>` element will be used.
+* If no `data-lang-default` is specified, "en" will be used as a last resort.
+
+### Resources Property Example
+
+In this example the language resources for `ar` and `en` are stored in the `__langResources` object.
+
+```javascript
+import { LocalizeMixin } from '@brightspace-ui/core/mixins/localize-mixin.js';
+class MyComponent extends LocalizeMixin(LitElement) {
+	static get properties() {
+		return {
+			__langResources: { type: Object }
+		};
+	}
+
+	constructor() {
+		super();
+		this.__langResources = {
+			'ar': {
+				more: 'المزيد',
+				less: 'أقل'
+			},
+			'en': {
+				more: 'more',
+				less: 'less'
+			}
+		};
+	}
+
+	/*
+	* Gets the current language based upon resources in this.__langResources property
+	* Checks each language in `langs` (e.g., ['ar-dz', 'ar', 'en']) and returns the first language with resources
+	*/
+	getLanguage(langs) {
+		for (let i = 0; i < langs.length; i++) {
+			if (this.__langResources[langs[i]]) {
+				return langs[i];
+			}
+		}
+
+		return null;
+	}
+
+	/*
+	* Checks if language resources have already been retrieved for the lang
+	* If so, returns the resources
+	* If not, gets the language resources for `lang` from this.__langResources object
+	*/
+	async getLangResources(lang) {
+		const proto = this.constructor.prototype;
+		this.checkLocalizationCache(proto);
+
+		const namespace = `more-less:${lang}`;
+
+		if (proto.__localizationCache.requests[namespace]) {
+			return proto.__localizationCache.requests[namespace];
```

Comment:
I don't think it's ideal for the sub-class to be calling into the base class's private properties here.

Since in this case (and the example below) it's checking a cache and if it's not there it's providing a way to fetch the value, maybe we could implement something in the base class that will do all that for us.

Then this class would just have to do something like this:

```javascript
async getLangResources(lang) {
    const namespace = `more-less:${lang}`;
    return this.localizationCache(namespace, () => {
        return this.__langResources[lang];
    });
}
```

And the dynamic example below:

```javascript
async getLangResources(lang) {
    const namespace = `more-less:${lang}`;
    return this.localizationCache(namespace, async () => {
        let translations;
        switch (lang) {
            case 'en':
                translations = await import('./locales/en.js');
                break;
            case 'ar':
                translations = await import('./locales/ar.js');
                break;
        }
        resolve(translations.val);
    });
}
```

So the underlying `localizationCache()` would:
- initialize the cache if it hasn't been already
- check if there's an entry for the key provided -- if so, return it
- otherwise it would call the function provided and put the result in the cache (which may be a promise)